cdap-security/src/main/java/co/cask/cdap/security/auth/context/MasterAuthenticationContext.java

```diff
@@ -19,17 +19,40 @@
 import co.cask.cdap.proto.security.Principal;
 import co.cask.cdap.security.spi.authentication.AuthenticationContext;
 import co.cask.cdap.security.spi.authentication.SecurityRequestContext;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.io.IOException;
 
 /**
  * An {@link AuthenticationContext} for HTTP requests in the Master. The authentication details in this context are
- * derived from {@link SecurityRequestContext}.
+ * derived from:
+ * <ol>
+ *   <li>{@link SecurityRequestContext}, when the request; or</li>
+ *   <li>{@link UserGroupInformation}, when the master itself is asynchronously updating privileges in the
+ *   authorization policy cache.</li>
+ * </ol>
  *
  * @see SecurityRequestContext
+ * @see UserGroupInformation
  */
 public class MasterAuthenticationContext implements AuthenticationContext {
 
   @Override
   public Principal getPrincipal() {
-    return SecurityRequestContext.toPrincipal();
+    // When requests come in via rest endpoints, the userId is updated inside SecurityRequestContext, so give that
+    // precedence.
+    String userId = SecurityRequestContext.getUserId();
+    // This userId can be null, when the master itself is asynchoronously updating the policy cache, since
+    // during that process the router will not set the SecurityRequestContext. In that case, obtain the userId from
+    // the UserGroupInformation, which will be the user that the master is running as.
+    if (userId == null) {
```

Comment:
When RemoteSystemOperationsService makes a call to AppFabric, SecurityRequestCOntext is not set, because it doesn't go via router. So have to use it. However, gave precedence to SecurityRequestContext so that in master REST APIs, we use SecurityRequestContext.